### ISSUE TITLE
Most composables can not be used in modal or other injected overlays (vue-demo-store)

### ISSUE THREAD

**Issue by benjamin-ott** [shopware/frontends#64]
Hi,
I'll try to explain my problem with a specific use case, but I'm sure there are many others.

I moved the listing filters in an overlay, but to my surprise the filters were not there anymore.
The reason for this should be the new useListing composable call. I think the inject in the composable is not working, because the overlay itself is higher up in the component tree (in the layout).

I also noticed the useAddress composable is called in the app.vue. This way the address modals in the account are working correctly because the highest provide is happening in the app.vue. If this global call for useAdress is removed it is not working anymore.

And the global call is not working for the useListing composable and anyways i think it's not a good idea to call every composable in the app which has an overlay somewhere.

Do you have any tipps for me how this problem could be solved? I think for most of the use cases an overlay is used to move functionality inside the overlay and so it is important that the composables can be used there.

**Comment by BrocksiNet** [comment]
Can you provide some example code? 

Or more file paths from where you removed what part of code to where and so on.

**Comment by benjamin-ott** [comment]
I tried to use stackblitz to build an example, but i sadly can't get the modal to pen there.. somehow the TestModal component is not rendered: https://stackblitz.com/edit/nuxt-starter-xes3df?file=pages/index.vue

But you can also test it in the vue-demo-store by removing the following line of code:
https://github.com/shopware/frontends/blob/main/templates/vue-demo-store/app.vue#L37

After that change please go to the account address page and make some changes via an address modal. The address on the page in the background won't update. If the useAddress composable is called in app.vue it is working again.

**Comment by patzick** [comment]
Hey @benjamin-ott 
Provide/inject mechanism is explained in the `context composables` section: https://shopware-frontends-docs.vercel.app/framework/context-composables.html

The mechanism allows us to share states between composables but also create new "contexts" and not have a single global state to handle. 
In the old way, we had `useSharedState` for that, but it was not flexible, and often you had a lot of unused data in memory, as you couldn't detect what was unnecessary anymore.
With Context Composables, you're defining the scope where you want to be reached. Imagine you'd like to display a CMS page and then create Quickview with another CMS page. Then you create a new Context, and it's treated as a root.
In the case of `useAddress` in the template, we added it there as it manages the user's address when he's logged in.

But please remember that the template is just a reference implementation and may not fit your specific needs. In this case, you should define on your own where the `root` of your scope is and where it should be placed. If you place it, for example, in Layout - you can share the scope between pages. If you put it into the page - you can use it there, and it's cleaned after the page is changed (very good for memory).

A good example is that you mentioned `useListing`. We keep the scope on the Page level, and we're initialising shared memory only on Navigation pages - https://github.com/shopware/frontends/blob/85639d96c2fdc96b1df091c30b4f2f778d88a83a/packages/cms-base/components/public/cms/CmsPage.vue#L12-L14
so the memory is cleaned with page change.
In conclusion - totally agree that you should not call every context composable in the root App, as the `provide` from this lasts for the whole application lifecycle. You must consider what's needed for you there and use `page` or even smaller scopes in other use cases. Let me know if that explains it well or how we could help here more :)

**Comment by benjamin-ott** [comment]
@patzick Hi!

Thank you for your explanation! It confirmed my understanding of the new structure.

The missing thing is in your last sentence.. I just didn't know where to place my modal component to make it work everywhere with the provide/inject mechanism.
I also tried Teleport, but i couldn't make it work. But if you think it could work i will try again and put a little bit more time into it. 
And if someone on your side has also the time to try it out with Teleport, i would really appreciate it 😃

**Comment by BrocksiNet** [comment]
@benjamin-ott Last friday I was playing with modals for you :heart: 
Feel free to test and provide feedback if you want. See this [PR](https://github.com/shopware/frontends/pull/148).

**Comment by benjamin-ott** [comment]
@BrocksiNet Thank you so much, this helped a lot!

I changed some things, but this was a really good base to work with.

- i used `v-show="visible"` instead of  `:class="[visible ? 'block' : 'hidden']"` because the transition was not working for me
- i added some logic to the modal provide because i wanted to supply the component which should be loaded as the modal content via the modal open method (like in the vue-demo-store template)
- i also added click.stop on the modal-content because the modal closed itself if i clicked on it
`<div
      id="modal-content"
      class="px-5 pb-7.5"
      @click.stop
  />`

**Comment by BrocksiNet** [comment]
Feel free to commit your improvements to the PR. Sounds great. :+1: 
Would be really nice.

**Comment by BrocksiNet** [comment]
Patryk did a great update to the modal example 🚀 

[Here](https://github.com/shopware/frontends/pull/165) we use [createSharedComposable](https://vueuse.org/shared/createSharedComposable/) instead of inject/provide.

Also we included the improvements from benjamin. 🍀

**Comment by benjamin-ott** [comment]
@BrocksiNet Thank you! Looks really good. From my point of view the issue can be closed.
